druid/examples/tabs.rs

```diff
@@ -0,0 +1,212 @@
+use druid::im::Vector;
+use druid::widget::{
+    Axis, Button, CrossAxisAlignment, Flex, Label, MainAxisAlignment, Padding, RadioGroup,
+    SizedBox, Split, TabInfo, Tabs, TabsPolicy, TabsTransition, TextBox, ViewSwitcher,
+};
+use druid::{theme, AppLauncher, Color, Data, Env, Lens, LensExt, Widget, WidgetExt, WindowDesc};
+use instant::Duration;
+
+#[derive(Data, Clone)]
+struct Basic {}
+
+#[derive(Data, Clone, Lens)]
+struct Advanced {
+    highest_tab: usize,
+    removed_tabs: usize,
+    tab_labels: Vector<usize>,
+}
+
+impl Advanced {
+    fn new(highest_tab: usize) -> Self {
+        Advanced {
+            highest_tab,
+            removed_tabs: 0,
+            tab_labels: (1..=highest_tab).collect(),
+        }
+    }
+
+    fn add_tab(&mut self) {
+        self.highest_tab += 1;
+        self.tab_labels.push_back(self.highest_tab);
+    }
+
+    fn remove_tab(&mut self, idx: usize) {
+        if idx >= self.tab_labels.len() {
+            log::warn!("Attempt to remove non existent tab at index {}", idx)
+        } else {
+            self.removed_tabs += 1;
+            self.tab_labels.remove(idx);
+        }
+    }
+
+    fn tabs_key(&self) -> (usize, usize) {
+        (self.highest_tab, self.removed_tabs)
+    }
+}
+
+#[derive(Data, Clone, Lens)]
+struct TabConfig {
+    axis: Axis,
+    cross: CrossAxisAlignment,
+    transition: TabsTransition,
+}
+
+#[derive(Data, Clone, Lens)]
+struct AppState {
+    tab_config: TabConfig,
+    basic: Basic,
+    advanced: Advanced,
+    text: String,
+}
+
+pub fn main() {
+    // describe the main window
+    let main_window = WindowDesc::new(build_root_widget)
+        .title("Tabs")
+        .window_size((700.0, 400.0));
+
+    // create the initial app state
+    let initial_state = AppState {
+        tab_config: TabConfig {
+            axis: Axis::Horizontal,
+            cross: CrossAxisAlignment::Start,
+            transition: Default::default(),
+        },
+        basic: Basic {},
+        advanced: Advanced::new(2),
+        text: "Interesting placeholder".into(),
+    };
+
+    // start the application
+    AppLauncher::with_window(main_window)
+        .use_simple_logger()
+        .launch(initial_state)
+        .expect("Failed to launch application");
+}
+
+fn build_root_widget() -> impl Widget<AppState> {
+    fn decor<T: Data>(label: Label<T>) -> SizedBox<T> {
+        label
+            .padding(5.)
```

Comment:
I'll check that, not sure I knew about it when I wrote this

Reply:
Trying it, this seems to increase the code noise of layout noise quite a bit, because factoring it out is harder (multiple calls onto a flex instead of a returned value ), meaning cut and paste is more likely.

Why is the spacer better than padding, or why does it matter? Also it isn't doing the same thing is it - padding is on all edges, and the spacer is only in the direction of the flex.